src/NUnitFramework/framework/Internal/AsyncToSyncAdapter.cs

```diff
@@ -0,0 +1,150 @@
+// ***********************************************************************
+// Copyright (c) 2013–2018 Charlie Poole, Rob Prouse
+//
+// Permission is hereby granted, free of charge, to any person obtaining
+// a copy of this software and associated documentation files (the
+// "Software"), to deal in the Software without restriction, including
+// without limitation the rights to use, copy, modify, merge, publish,
+// distribute, sublicense, and/or sell copies of the Software, and to
+// permit persons to whom the Software is furnished to do so, subject to
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+// ***********************************************************************
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Security;
+using System.Threading;
+using NUnit.Compatibility;
+
+namespace NUnit.Framework.Internal
+{
+    internal static class AsyncToSyncAdapter
+    {
+        public static bool IsAsyncOperation(MethodInfo method)
+        {
+            return IsTaskType(method.ReturnType) ||
+                   method.GetCustomAttributes(false).Any(attr => attr.GetType().FullName == "System.Runtime.CompilerServices.AsyncStateMachineAttribute");
+        }
+
+        public static bool IsAsyncOperation(Delegate @delegate)
+        {
+            return IsAsyncOperation(@delegate.GetMethodInfo());
+        }
+
+        private static bool IsTaskType(Type type)
+        {
+            for (; type != null; type = type.GetTypeInfo().BaseType)
+            {
+                if (type.GetTypeInfo().IsGenericType
+                    ? type.GetGenericTypeDefinition().FullName == "System.Threading.Tasks.Task`1"
+                    : type.FullName == "System.Threading.Tasks.Task")
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+#if ASYNC
+        private const string TaskWaitMethod = "Wait";
+        private const string TaskResultProperty = "Result";
+        private const string VoidTaskResultType = "VoidTaskResult";
+        private const string SystemAggregateException = "System.AggregateException";
+        private const string InnerExceptionsProperty = "InnerExceptions";
```

Comment:
Are these strings mostly just used the once? Is there a reason not to inline them?

Reply:
This code was [moved](https://github.com/nunit/nunit/pull/2818/files/4be14d688bb6e5948157062a58eb4e2db2eebbb3#diff-c4545d57570ce4b1269a9f0adf8a4b18L99) without changing it.
In the next PR when I add support for general C# awaitables (and maybe F# too), I'll be pulling this stuff out anyway.

Reply:
Spot on. I'm going to leave it since this code was [moved](https://github.com/nunit/nunit/pull/2818/files/4be14d688bb6e5948157062a58eb4e2db2eebbb3#diff-c4545d57570ce4b1269a9f0adf8a4b18L99) without changing it, and in the next PR when I add support for general C# awaitables (and maybe F# too), I'll be pulling this stuff out anyway.